j2cl-tasks/src/main/java/com/vertispan/j2cl/build/provided/TranslationsFileProcessor.java

```diff
@@ -97,17 +98,109 @@ public Optional<File> getTranslationsFile(List<Input> inputs, BuildLog log) {
 
                     String lang = translationbundleNode.item(0).getAttributes().getNamedItem("lang").getNodeValue();
 
-                    if (locale.equals(lang)) {
-                        return Optional.of(xtb);
+                    if (locales.contains(lang)) {
+                        if(!suitableFiles.containsKey(lang)) {
+                            suitableFiles.put(lang, new HashSet<>());
+                        }
+
+                        suitableFiles.get(lang).add(translationbundleNode);
                     }
                 }
+                if(!suitableFiles.isEmpty()) {
+                    return mergeFiles(suitableFiles, locales, context);
+                }
             } catch (ParserConfigurationException | SAXException | IOException e) {
-                log.error("Error while reading xtb files ", e);
+                context.log().error("Error while reading xtb files ", e);
                 throw new RuntimeException(e);
             }
-            log.warn("No matching locales for " + locale);
+            context.log().warn("No matching locales for " + locale);
             return Optional.empty();
         }
 
+
+        private Set<String> locales(String locale) {
+            Set<String> result = new HashSet<>();
+            result.add(locale);
+
+            if (locale.contains("_")) {
+                // according to spec, locale discaration can have 3 parts max
+                if(locale.indexOf("_") != locale.lastIndexOf("_")) {
+                    result.add(locale.substring(0, locale.lastIndexOf("_")));
+                }
+                result.add(locale.substring(0, locale.indexOf("_")));
+            }
+            return result;
+        }
+
+        private Optional<File> mergeFiles(HashMap<String, Set<NodeList>> suitableFiles, Set<String> locales, TaskContext context) {
+            Map<String, Node> resultedCodeSet = new HashMap<>();
+            // parent nodes first
+            locales.stream().sorted((o1, o2) -> {
+                if(o1.split("_").length > o2.split("_").length) {
+                    return 1;
+                }
+                return -1;
+            }).forEach(locale -> {
+                suitableFiles.get(locale).forEach(node -> {
+                    NodeList children = node.item(0).getChildNodes();
+
+                    for (int i = 0; i < children.getLength(); i++) {
+                        if (children.item(i).getNodeName().equals("translation")) {
+                            String key = children.item(i).getAttributes().getNamedItem("id").getNodeValue();
+                            // override parent keys if already exists
+                            resultedCodeSet.put(key, children.item(i));
```

Comment:
take the result of this and warn/error if one file overwrites another if they have matching IDs?

Reply:
I might have misunderstood, i thought that `id` was the key for the translation entry. If two libraries in a project accidentally set the same id and so overwrite each other, that's bad, right?

Reply:
in my case, child overrides parent, so it's ok